### PR TITLE
Support custom refspec in lightweight checkout

### PR DESCRIPTION
**Pull request by ckullabosch**
Let user specify a custom refspec to fetch.
Add support for branch names FETCH_HEAD or a commit hash.

This allows to use lightweight checkout for:

* Fetch the pipeline from a Gerrit change (refspec:
  refs/changes/*/change/patchset and branch: FETCH_HEAD).

* Fetch the pipeline from a fixed commit of a branch, instead of head (refspec: branch and branch: commit hash).

## [JENKINS-xxxxx](https://issues.jenkins.io/browse/JENKINS-xxxxx) - summarize pull request in one line

Describe the big picture of your changes here to explain to the maintainers why we should accept this pull request.
If it fixes a bug or resolves a feature request, include a link to the issue.

## Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. This is simply a reminder of what we are going to look for before merging your code. If a checkbox or line does not apply to this pull request, delete it. We prefer all checkboxes to be checked before a pull request is merged_

- [ ] I have read the [CONTRIBUTING](https://github.com/jenkinsci/git-plugin/blob/master/CONTRIBUTING.adoc) doc
- [ ] I have referenced the Jira issue related to my changes in one or more commit messages
- [ ] I have added tests that verify my changes
- [ ] Unit tests pass locally with my changes
- [ ] I have added documentation as necessary
- [ ] No Javadoc warnings were introduced with my changes
- [ ] No spotbugs warnings were introduced with my changes
- [ ] Documentation in README has been updated as necessary
- [ ] Online help has been added and reviewed for any new or modified fields
- [ ] I have interactively tested my changes
- [ ] Any dependent changes have been merged and published in upstream modules (like git-client-plugin)

## Types of changes

- [x] New feature (non-breaking change which adds functionality)

### CODE REVIEW

**Comment by ckullabosch** [comment]
@MarkEWaite I think this PR is ready for review.

**Comment by MarkEWaite** [comment]
> @MarkEWaite I think this PR is ready for review. Please let me know if I can do anything to drive the topic forward.

Thanks @ckullabosch .  It will likely be a week before I'm able to review it due to other priorities.  Weekend is coming that may make some time available.

**Comment by MarkEWaite** [comment]
It would be a great help if you could provide one or more examples of how the custom refespec can be used.  The most frequent request from Jenkins documentation readers is a request for more examples.  The README file of the git plugin provides many examples in hopes that readers will find them useful.  The documentation that you create will also be used as part of my testing of the pull request, so it will help me evaluate the change more effectively.

**Review by MarkEWaite — CHANGES_REQUESTED** [review]
I'd really like to have documentation included that provides an example using the new capabilities, either through a Pipeline example (if that applies) or through a configuration as code example.

I still need to perform interactive testing.  My review of the tests and the code has not found any blocking issues, just questions that I need to answer when testing.

**Comment by MarkEWaite** [comment]
With the additional spotbugs checks that were added in #1475 , there is now a spotbugs warning reported for the unread field `HeadNameResult.headName`.  The report seems to be correct because the value of headName that is passed to the `HeadNameResult` constructor is ignored and replaced with a value computed from the branchSpec.

**Comment by MarkEWaite** [comment]
@ckullabosch I'll need more detailed instructions on the steps that I need to take in order to verify this works as expected.  I have been unable to reach the modified code from my job definitions.  I created a Pipeline, a multibranch Pipeline, and an organization folder and was unable in any of those cases to see the "Done with " message logged.

Can you provide more details of a way that I can verify the code using GitHub or Bitbucket or GitLab or Gitea?  I don't have access to a Gerrit server.

File: src/main/java/jenkins/plugins/git/GitSCMFileSystem.java [hunk 1]
```diff
@@ -287,39 +290,76 @@ public boolean supportsDescriptor(SCMSourceDescriptor descriptor) {
...
                     } else if (branchSpecExpandedName.startsWith("*/")) {
                         headName = branchSpecExpandedName.substring(2);
-                    } else {
-                        headName = branchSpecExpandedName;
                     }
                 }
-                return new HeadNameResult(headName, prefix);
+
+                if (refspecExpandedName == null || refspecExpandedName.equals("")) {
+                    refspecExpandedName = "+" + prefix + headName + ":" + Constants.R_REMOTES + remoteName + "/" + headName;
```
[the remark this answers is not in the document]
Comment (ckullabosch) [reply src/main/java/jenkins/plugins/git/GitSCMFileSystem.java]: Thanks for pointing out, fixed now

File: src/main/java/jenkins/plugins/git/GitSCMFileSystem.java [hunk 2]
```diff
@@ -287,39 +290,76 @@ public boolean supportsDescriptor(SCMSourceDescriptor descriptor) {
...
-                this.prefix = prefix;
+                this.remoteHeadName = remoteHeadName;
+                this.refspec = refspec;
+                this.rev = rev;
             }
 
             static HeadNameResult calculate(@NonNull BranchSpec branchSpec,
                                             @CheckForNull SCMRevision rev,
-                                            @CheckForNull EnvVars env) {
+                                            @NonNull String refSpec,
```
[the remark this answers is not in the document]
Comment (ckullabosch) [reply src/main/java/jenkins/plugins/git/GitSCMFileSystem.java]: refSpec should have the  @CheckForNull annotation, fixed now

**Comment by ckullabosch** [comment]
@MarkEWaite In order to try out the change you have to create a Pipeline job which loads the pipeline from SCM (git). You have to enable lightweight checkout (set the checkmark at the end of the job configuration page). Then you may want to experiment with different settings for refspec and branch. There is no need for a folder or multibranch project.

Settings which are possible now, but weren't before:

1. Have a refspec pointing to a branch, point the branch setting to a commit hash (not necessarily the tip of the branch). This lets you use a fixed commit which is not a ref, but contained in the history of a branch.
2. Have a refspec other than refs/tags and refs/heads and make use of "FETCH_HEAD" in the branch setting. The available refs depends on your git system (a branch like refs/heads/my-branch should do as well). In Gerrit you have the refs/changes/, probably other git systems have also special refs (I remember Bitbucket having refs for pull requests).

We need to clarify about improving the README. The lightweight checkout is not mentioned at all at the moment.

**Comment by MarkEWaite** [comment]
It will be 1-2 weeks before I can review and test this.  We're in the last few weeks of Google Summer of Code and I put my efforts there.

**Comment by ckullabosch** [comment]
@MarkEWaite It's been a while, any chance we can conclude this PR? Please see latest comment of mine.

> We need to clarify about improving the README. The lightweight checkout is not mentioned at all at the moment.

**Comment by SinOverCos** [comment]
@ckullabosch thank you for the PR, I've been looking for something like this.

@MarkEWaite any chance you could give this another pass?

**Comment by MarkEWaite** [comment]
> @MarkEWaite any chance you could give this another pass?

As time allows, I'll do another review.  @SinOverCos have you deployed the build of this change into your environment and confirmed that it works for you?  It is much more persuasive for a maintainer like me when others users declare that they are using a pull request in production.

File: src/main/java/jenkins/plugins/git/GitSCMFileSystem.java [hunk 3]
```diff
@@ -286,40 +289,80 @@ public boolean supportsDescriptor(SCMSourceDescriptor descriptor) {
...
 
+                // default to a branch (refs/heads)
                 String prefix = Constants.R_HEADS;
+                // check for a tag
                 if (branchSpecExpandedName.startsWith(Constants.R_TAGS)) {
                     prefix = Constants.R_TAGS;
+                } else {
+                    // check for FETCH_HEAD
+                    if (branchSpecExpandedName.equals(Constants.FETCH_HEAD) && refspecExpandedName != null &&
+                            !refspecExpandedName.equals("")) {
```
Comment (balakine) [inline src/main/java/jenkins/plugins/git/GitSCMFileSystem.java:325]: We already have `StringUtils` imported, might as well use it
```suggestion
                    if (branchSpecExpandedName.equals(Constants.FETCH_HEAD) && StringUtils.isNotBlank(refSpecExpandedName)) {
```

File: src/main/java/jenkins/plugins/git/GitSCMFileSystem.java [hunk 4]
```diff
@@ -286,40 +289,80 @@ public boolean supportsDescriptor(SCMSourceDescriptor descriptor) {
...
+                        prefix = null;
+                    } else {
+                        // check for commit-id
+                        final String regex = "^[a-fA-F0-9]{40}$";
+                        final Pattern pattern = Pattern.compile(regex, Pattern.MULTILINE);
+                        final Matcher matcher = pattern.matcher(branchSpecExpandedName);
+
+                        if (matcher.find()) {
```
Comment (balakine) [inline src/main/java/jenkins/plugins/git/GitSCMFileSystem.java:333]: There is an existing pattern for this [here](https://github.com/jenkinsci/git-plugin/blob/master/src/main/java/hudson/plugins/git/util/DefaultBuildChooser.java#L58C28-L58C64)
```suggestion
                        if (branchSpecExpandedName.matches("[0-9a-f]{6,40}")) {
```

File: src/main/java/jenkins/plugins/git/GitSCMFileSystem.java [hunk 5]
```diff
@@ -286,40 +289,80 @@ public boolean supportsDescriptor(SCMSourceDescriptor descriptor) {
...
-                        headName = branchSpecExpandedName.substring(prefix.length());
+                    if (prefix != null && branchSpecExpandedName.startsWith(prefix)) {
+                        calculatedHeadName = branchSpecExpandedName.substring(prefix.length());
                     } else if (branchSpecExpandedName.startsWith("*/")) {
-                        headName = branchSpecExpandedName.substring(2);
+                        calculatedHeadName = branchSpecExpandedName.substring(2);
+                    }
+                }
+
+                if (refspecExpandedName == null || refspecExpandedName.equals("")) {
```
Comment (balakine) [inline src/main/java/jenkins/plugins/git/GitSCMFileSystem.java:352]: `StringUtils` again
```suggestion
                if (StringUtils.isBlank(refspecExpandedName)) {
```

File: src/main/java/jenkins/plugins/git/GitSCMFileSystem.java [hunk 6]
```diff
@@ -286,40 +289,80 @@ public boolean supportsDescriptor(SCMSourceDescriptor descriptor) {
...
+                    if (prefix.equals(Constants.R_TAGS)) {
```
Comment (balakine) [inline src/main/java/jenkins/plugins/git/GitSCMFileSystem.java:353]: I'd suggest adding a test with a sha1 and a `null` refspec. This `if` would need to be fixed
```suggestion
                    if (prefix != null && prefix.equals(Constants.R_TAGS)) {
```

File: src/main/java/jenkins/plugins/git/GitSCMFileSystem.java [hunk 7]
```diff
@@ -286,40 +289,80 @@ public boolean supportsDescriptor(SCMSourceDescriptor descriptor) {
...
+                    } else {
+                        // check for commit-id
+                        final String regex = "^[a-fA-F0-9]{40}$";
+                        final Pattern pattern = Pattern.compile(regex, Pattern.MULTILINE);
+                        final Matcher matcher = pattern.matcher(branchSpecExpandedName);
+
+                        if (matcher.find()) {
+                            // commit-id
+                            prefix = null;
+                            rev = new AbstractGitSCMSource.SCMRevisionImpl(new SCMHead(branchSpecExpandedName), branchSpecExpandedName);
```
Comment (balakine) [inline src/main/java/jenkins/plugins/git/GitSCMFileSystem.java:336]: Overriding arguments like that deserves at least a comment.

File: src/main/java/jenkins/plugins/git/GitSCMFileSystem.java [hunk 8]
```diff
@@ -286,40 +289,80 @@ public boolean supportsDescriptor(SCMSourceDescriptor descriptor) {
...
+                        }
+                    }
                 }
 
-                String headName;
+                String calculatedHeadName = branchSpecExpandedName;
```
Comment (balakine) [inline src/main/java/jenkins/plugins/git/GitSCMFileSystem.java:341]: We don't _need_ to rename this variable.

**Review by balakine — COMMENTED** [review]
I really want to see this in merged. Perhaps reducing the line count would make it easier to review.

**Comment by monfortt** [comment]
Hello,
Is there any willing to merge this PR?
It contains a very valuable feature for us, so that lightweight checkouts support commit hashes.
And sadly, the PR is stuck for more than a year, which is a pity.
Unless another change recently delivered (after 5.2.2) provides the feature I mentioned?
Thank you in advance!

**Comment by MarkEWaite** [comment]
@monfortt  have you deployed the build of this change into your environment and confirmed that it works for you? It is much more persuasive for a maintainer like me when other users declare that they are using a pull request in production.

If the conflict were resolved and an incremental build were available, would you deploy it and confirm it works for you?
